lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/SharedState.scala

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark.lightgbm
+
+import java.io.{IOException, ObjectInputStream}
+import java.util.concurrent.CountDownLatch
+
+import com.microsoft.ml.spark.lightgbm.dataset.DatasetAggregator
+import com.microsoft.ml.spark.lightgbm.params.TrainParams
+import org.apache.spark.sql.types.StructType
+import org.slf4j.Logger
+
+class SharedState extends Serializable {
+  var mainExecutorWorker: Long = LightGBMUtils.getTaskId()
+
+  @volatile var datasetAggregator: Option[DatasetAggregator] = None
+  def linkDatasetAggregator(columnParams: ColumnParams,
+                            schema: StructType,
+                            trainParams: TrainParams): Unit = {
+    if (datasetAggregator.isEmpty) {
+      this.synchronized {
+        if (datasetAggregator.isEmpty) {
+          datasetAggregator =
+            Some(new DatasetAggregator(columnParams, schema, trainParams.executionParams.useSingleDatasetMode,
+              trainParams.executionParams.chunkSize, trainParams.executionParams.matrixType, this))
+        }
+      }
+    }
+  }
+
+  @volatile var isSparse: Option[Boolean] = None
+  def linkIsSparse(isSparse: Boolean): Unit = {
+    if (this.isSparse.isEmpty) {
+      this.synchronized {
+        if (this.isSparse.isEmpty) {
+          this.isSparse = Some(isSparse)
+        }
+      }
+    }
+  }
+
+  @transient @volatile var arrayProcessedSignal: CountDownLatch = new CountDownLatch(0)
```

Comment:
Yes, I think it needs to be volatile since it's updated and read by different threads.  Using volatile doesn't hurt much, and it seems safer to have it in this case.  All it means is that we need to go to main memory (RAM) instead of the L2/L3 cache on the processor.  If we are getting it from the cache on different threads that could cause bad issues.  If you are certain that it won't be retrieved from the thread's cache I can remove it though.